pkg/network/handlers/work_package_sharing_handler.go

```diff
@@ -0,0 +1,152 @@
+		return fmt.Errorf("failed to unmarshal share header: %w", err)
+	}
+
+	var rootMappings []SegmentRootMapping
+	if err = jam.Unmarshal(msg1.Content[2:], &rootMappings); err != nil {
+		return fmt.Errorf("failed to unmarshal work package: %w", err)
+	}
+
+	// Read the work-package bundle
+	msg2, err := ReadMessageWithContext(ctx, stream)
+	if err != nil {
+		return fmt.Errorf("failed to read work package bundle: %w", err)
+	}
+
+	var bundle work.PackageBundle
+	if err := jam.Unmarshal(msg2.Content, &bundle); err != nil {
+		return fmt.Errorf("failed to unmarshal work package bundle: %w", err)
+	}
+
+	log.Printf("Received work package bundle with %d work item(s) and %d extrinsic byte(s)",
+		len(bundle.Package.WorkItems), len(bundle.Extrinsics))
+
+	segmentRootLookup := buildSegmentRootLookup(rootMappings)
+
+	if err := h.verifySegmentRootMappings(segmentRootLookup, bundle); err != nil {
+		return fmt.Errorf("mappings verification failed: %w", err)
+	}
+
+	authOutput, err := h.auth.InvokePVM(bundle.Package, coreIndex)
+	if err != nil {
+		return fmt.Errorf("authorization failed: %w", err)
+	}
+
+	workReport, err := ProduceWorkReport(ctx, h.refine, h.serviceState, authOutput, coreIndex, bundle, segmentRootLookup)
+	if err != nil {
+		return fmt.Errorf("failed to produce work report: %w", err)
+	}
+
+	workReportHash, err := workReport.Hash()
+	if err != nil {
+		return fmt.Errorf("failed to hash work report: %w", err)
+	}
+
+	signature := ed25519.Sign(h.privateKey, workReportHash[:])
+
+	// Prepare the response: Work-Report Hash ++ Ed25519 Signature.
+	response := struct {
+		WorkReportHash crypto.Hash
+		Signature      []byte
+	}{
+		WorkReportHash: workReportHash,
+		Signature:      signature,
+	}
+
+	respBytes, err := jam.Marshal(response)
+	if err != nil {
+		return fmt.Errorf("failed to marshal response: %w", err)
+	}
+
+	if err := WriteMessageWithContext(ctx, stream, respBytes); err != nil {
+		return fmt.Errorf("failed to write response: %w", err)
+	}
+
+	if err := stream.Close(); err != nil {
+		return fmt.Errorf("failed to close stream: %w", err)
+	}
+
+	log.Printf("CE-134: Sent work-report hash and signature successfully")
+
+	return nil
+}
+
+func buildSegmentRootLookup(mappings []SegmentRootMapping) map[crypto.Hash]crypto.Hash {
+	lookup := make(map[crypto.Hash]crypto.Hash)
+	for _, m := range mappings {
+		lookup[m.WorkPackageHash] = m.SegmentRoot
+	}
+	return lookup
+}
+
+// simple verification (wp hash -> segment root)
+func (h *WorkPackageSharingHandler) verifySegmentRootMappings(lookup map[crypto.Hash]crypto.Hash, bundle work.PackageBundle) error {
+	for _, item := range bundle.Package.WorkItems {
+		for _, imp := range item.ImportedSegments {
+			if _, exists := lookup[imp.Hash]; !exists {
+				log.Printf("Missing mapping for imported segment hash %x", imp.Hash)
+				return fmt.Errorf("missing segment-root mapping for hash: %x", imp.Hash)
+			}
+		}
+	}
+
+	log.Print("Segment-root mappings verified")
```

Comment:
is this meant to be here?

Reply:
Yes I have put logs in various places for tracking the workflow.